hpx/lcos/promise.hpp

```diff
@@ -167,97 +164,91 @@ namespace lcos {
-        // Error conditions:
-        //   - future_already_retrieved if get_shared_future has already been
-        //     called on a promise with the same shared state as *this.
-        //   - no_state if *this has no shared state.
+        /// \returns   A shared_future<Result> object with the same shared state
+        ///            as *this.
+        /// \throws    future_error if *this has no shared state or if
+        ///            get_shared_future has already been called on a promise
+        ///            with the same shared state as *this.
+        ///            future_already_retrieved if get_shared_future has already been
+        ///            called on a promise with the same shared state as *this.
+        ///            no_state if *this has no shared state.
         using base_type::get_shared_future;
 
-        // Effects: atomically stores the value r in the shared state and makes
-        //          that state ready (30.6.4).
-        // Throws:
-        //   - future_error if its shared state already has a stored value or
-        //     exception, or
-        //   - any exception thrown by the constructor selected to copy an
-        //     object of R.
-        // Error conditions:
-        //   - promise_already_satisfied if its shared state already has a
-        //     stored value or exception.
-        //   - no_state if *this has no shared state.
+        /// \brief     atomically stores the value r in the shared state and makes
+        ///            that state ready (30.6.4).
+        /// \throws    future_error if its shared state already has a stored value or
+        ///            exception, or any exception thrown by the constructor
+				///            selected to copy an object of R.
+        ///            promise_already_satisfied if its shared state already has a
+        ///            stored value or exception.
+        ///            no_state if *this has no shared state.
         void set_value()
         {
             base_type::set_value(hpx::util::unused);
         }
 
-        // Effects: atomically stores the exception pointer p in the shared
-        //          state and makes that state ready (30.6.4).
-        // Throws: future_error if its shared state already has a stored value
-        //         or exception.
-        // Error conditions:
-        //   - promise_already_satisfied if its shared state already has a
-        //     stored value or exception.
-        //   - no_state if *this has no shared state.
+        /// \brief     atomically stores the exception pointer p in the shared
+        ///            state and makes that state ready (30.6.4).
+        /// \throws    future_error if its shared state already has a stored value
+        ///            or exception.
+        ///            promise_already_satisfied if its shared state already has a
+        ///            stored value or exception.
+        ///            no_state if *this has no shared state.
```

Comment:
This is the only part that still needs some work. The parts that were under "Error conditions" need better formatting (see [here](https://en.cppreference.com/w/cpp/thread/promise/set_value) under "Exceptions" for an example). `future_error` is the type of the exception, whereas `promise_already_satisfied` and `no_state` are error codes associated with the exception. So maybe change it to something like (ignore the ugly formatting):
```
/// \throws future_error if [... as before ...] exception. The error code
/// is set to promise_already_satisfied if its shared state...
/// The error code is set to no_state if...
```
The same goes for all the other "Error conditions" sections in this file. I think after that we should be able to merge this. Thanks for working on this!

Reply:
What I meant was to update them to look like this: https://github.com/STEllAR-GROUP/hpx/blob/8b6c014fe5aa2f6e507ac48065006bca20835e84/hpx/lcos/promise.hpp#L232-L238. I liked that change (except that it shouldn't say `[... as before ...]`).

Reply:
Cool, will do.

Reply:
Sure.

Reply:
@msimberg I still don't get you, wdym?

Reply:
Sorry if I was unclear. Having both `\throws` and `\exception` doesn't make much sense. I'd make it one `\throws` section which explains the two different error codes that can be set.

Easiest might be to look at https://en.cppreference.com/w/cpp/thread/promise/set_value and use the documentation from there.

The same change should be made for all the other member functions. Does that make it clearer?

Reply:
I can see what you are saying. I can remove that, np.

Reply:
I have another tiny input: You can have a `\error` tag as well, should I add that or would you just prefer adding `\throws` and `\exception` wherever strictly required?

Reply:
As far as I can tell `\error` is not a built-in Doxygen command: http://www.doxygen.nl/manual/commands.html#cmdxrefitem. Do you have another source for this? I think we'll be okay without `\error`.

Reply:
Ok.